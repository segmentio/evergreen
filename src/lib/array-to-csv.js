import isEmpty from 'lodash.isempty'

/**
 * Converts an array to a comma separated string
 * @param {string[]} value
 * @returns {string}
 */
const arrayToCsv = value => {
  // Accept only array values + ensure there values inside the array, otherwise
  // there's nothing to join
  if (!Array.isArray(value) || isEmpty(value)) {
    return ''
  }

  return value.join(',')
}

export default arrayToCsv
