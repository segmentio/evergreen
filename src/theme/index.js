export { default as ThemeContext, ThemeProvider, ThemeConsumer } from './src/ThemeContext'
export { default as withTheme } from './src/withTheme'
export { default as useTheme } from './src/useTheme'
export { get, getValue, resolveThemeTokens } from './src/theme-tools'
