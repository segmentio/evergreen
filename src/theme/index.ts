export { default as ThemeContext, getThemeContext, ThemeProvider, ThemeConsumer } from './src/ThemeContext'
export { default as withTheme } from './src/withTheme'
export { default as useTheme } from './src/useTheme'
export { get, getValue, mergeTheme, resolveThemeTokens } from './src/theme-tools'
