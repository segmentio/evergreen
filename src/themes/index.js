export { default as defaultTheme } from './default'
