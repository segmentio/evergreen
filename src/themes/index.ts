export { default as defaultTheme } from './default'
export type {
  Components,
  DefaultTheme,
  DefaultThemeAppearances,
  DefaultThemeColors,
  DefaultThemeColor,
  DefaultThemeFills,
  DefaultThemeIntents,
  DefaultThemeIntent,
  DefaultThemePseudoSelectors,
  DefaultThemeSizes,
} from './default'
export { default as classicTheme } from './classic'
export { default as deprecatedDefaultTheme } from './deprecated'
