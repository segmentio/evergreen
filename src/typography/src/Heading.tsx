import React, { forwardRef, memo } from 'react'
import cx from 'classnames'
import PropTypes from 'prop-types'
import Box from 'ui-box'
import { useStyleConfig } from '../../hooks'

const pseudoSelectors = {}
const internalStyles = {}

const Heading = memo(
  forwardRef(function Heading(props, ref) {
    const { className, size = 500, ...restProps } = props
    const { className: themedClassName, ...styleProps } = useStyleConfig(
      'Heading',
      { size },
      pseudoSelectors,
      internalStyles
    )

    return (
      <Box
        is="h2"
        ref={ref}
        className={cx(themedClassName, className)}
        marginTop={0}
        marginBottom={0}
        {...styleProps}
        {...restProps}
      />
    )
  })
)

Heading.propTypes = {
  /**
   * Heading composes Box as the base.
   */
  ...Box.propTypes,

  /**
   * The size of the heading.
   */
  size: PropTypes.oneOf([100, 200, 300, 400, 500, 600, 700, 800, 900])
}

export default Heading
