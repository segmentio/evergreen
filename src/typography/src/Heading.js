import React, { memo, forwardRef } from 'react'
import PropTypes from 'prop-types'
import Box from 'ui-box'
import { useTheme } from '../../theme'

const Heading = memo(
  forwardRef((props, ref) => {
    const theme = useTheme()
    const { marginTop, size, ...restProps } = props
    const {
      marginTop: defaultMarginTop,
      ...headingStyle
    } = theme.getHeadingStyle(size)

    let finalMarginTop = marginTop
    if (marginTop === 'default') {
      finalMarginTop = defaultMarginTop
    }

    return (
      <Box
        is="h2"
        innerRef={ref}
        marginTop={finalMarginTop || 0}
        marginBottom={0}
        {...headingStyle}
        {...restProps}
      />
    )
  })
)

Heading.propTypes = {
  /**
   * Heading composes Box as the base.
   */
  ...Box.propTypes,

  /**
   * The size of the heading.
   */
  size: PropTypes.oneOf([100, 200, 300, 400, 500, 600, 700, 800, 900]),

  /**
   * Pass `default` to use the default margin top for that size.
   */
  marginTop: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.number,
    PropTypes.string
  ])
}

Heading.defaultProps = {
  size: 500
}

export default Heading
