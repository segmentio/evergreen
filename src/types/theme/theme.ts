import { Components } from '../../themes'
import { Color } from './color'
import { ComponentStyles } from './component-styles'
import { Fill } from './fill'
import { FontFamilies } from './font-families'
import { FontWeights } from './font-weights'
import { Intent } from './intent'
import { LettingSpacings } from './letter-spacings'
import { ZIndices } from './z-indices'

export interface Theme<TComponents extends Components = Components> {
  colors: { [color: string]: Color<string | { [group: string]: Color }> }
  fills: { [fill: string]: Fill }
  intents: { [intent: string]: Intent }
  fontFamilies: FontFamilies
  radii: string[]
  shadows: string[] & { focusRing: string }
  fontSizes: string[]
  fontWeights: FontWeights
  letterSpacings: LettingSpacings
  lineHeights: string[]
  zIndices: ZIndices
  components: Partial<ComponentStyles<TComponents>>
}
