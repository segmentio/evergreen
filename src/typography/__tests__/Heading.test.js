import React from 'react'
import { render, screen } from '@testing-library/react'
import renderer from 'react-test-renderer'
import { UIBoxSerializer } from '../../../lib/testing'
import { ThemeProvider } from '../../theme'
import { defaultTheme } from '../../themes'
import Heading from '../src/Heading'

expect.addSnapshotSerializer(UIBoxSerializer)

test.each([
  ['size 100', 100],
  ['size 200', 200],
  ['size 300', 300],
  ['size 400', 400],
  ['size 500', 500],
  ['size 600', 600],
  ['size 700', 700],
  ['size 800', 800],
  ['size 900', 900]
])('<Heading /> %s renders as expected', (_, size) => {
  const component = (
    <ThemeProvider value={defaultTheme}>
      <Heading size={size}>{`Heading ${size}`}</Heading>
    </ThemeProvider>
  )
  const tree = renderer.create(component).toJSON()
  expect(tree).toMatchSnapshot()
})

test('Heading lets you override the underlying DOM element', async () => {
  render(
    <ThemeProvider value={defaultTheme}>
      <Heading is="h1">Testing h1</Heading>
    </ThemeProvider>
  )

  expect(screen.getByText('Testing h1', { selector: 'h1' })).toBeTruthy()
})
