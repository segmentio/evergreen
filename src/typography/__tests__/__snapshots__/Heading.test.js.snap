// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Heading /> size 100 renders as expected 1`] = `
Extracted Styles:
box-sizing: border-box;
color: #696f8c;
font-family: "SF UI Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 10px;
font-weight: 500;
letter-spacing: 0.6px;
line-height: 16px;
margin-bottom: 0px;
margin-top: 0px;
text-transform: uppercase;


<h2
  className="ub-mt_0px ub-mb_0px ub-color_696f8c ub-fnt-fam_b77syt ub-f-wght_500 ub-fnt-sze_10px ub-txt-trns_uppercase ub-ln-ht_16px ub-ltr-spc_0-6px ub-box-szg_border-box"
>
  Heading 100
</h2>
`;

exports[`<Heading /> size 200 renders as expected 1`] = `
Extracted Styles:
box-sizing: border-box;
color: #696f8c;
font-family: "SF UI Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 12px;
font-weight: 500;
letter-spacing: 0;
line-height: 16px;
margin-bottom: 0px;
margin-top: 0px;


<h2
  className="ub-mt_0px ub-mb_0px ub-color_696f8c ub-fnt-fam_b77syt ub-f-wght_500 ub-fnt-sze_12px ub-ln-ht_16px ub-ltr-spc_0 ub-box-szg_border-box"
>
  Heading 200
</h2>
`;

exports[`<Heading /> size 300 renders as expected 1`] = `
Extracted Styles:
box-sizing: border-box;
color: #101840;
font-family: "SF UI Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 12px;
font-weight: 500;
letter-spacing: 0;
line-height: 16px;
margin-bottom: 0px;
margin-top: 0px;


<h2
  className="ub-mt_0px ub-mb_0px ub-color_101840 ub-fnt-fam_b77syt ub-f-wght_500 ub-fnt-sze_12px ub-ln-ht_16px ub-ltr-spc_0 ub-box-szg_border-box"
>
  Heading 300
</h2>
`;

exports[`<Heading /> size 400 renders as expected 1`] = `
Extracted Styles:
box-sizing: border-box;
color: #101840;
font-family: "SF UI Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 14px;
font-weight: 500;
letter-spacing: -0.05px;
line-height: 18px;
margin-bottom: 0px;
margin-top: 0px;


<h2
  className="ub-mt_0px ub-mb_0px ub-color_101840 ub-fnt-fam_b77syt ub-f-wght_500 ub-fnt-sze_14px ub-ln-ht_18px ub-ltr-spc_-0-05px ub-box-szg_border-box"
>
  Heading 400
</h2>
`;

exports[`<Heading /> size 500 renders as expected 1`] = `
Extracted Styles:
box-sizing: border-box;
color: #101840;
font-family: "SF UI Text", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 16px;
font-weight: 600;
letter-spacing: -0.05px;
line-height: 20px;
margin-bottom: 0px;
margin-top: 0px;


<h2
  className="ub-mt_0px ub-mb_0px ub-color_101840 ub-fnt-fam_b77syt ub-f-wght_600 ub-fnt-sze_16px ub-ltr-spc_-0-05px ub-ln-ht_20px ub-box-szg_border-box"
>
  Heading 500
</h2>
`;

exports[`<Heading /> size 600 renders as expected 1`] = `
Extracted Styles:
box-sizing: border-box;
color: #101840;
font-family: "SF UI Display", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 18px;
font-weight: 600;
letter-spacing: -0.07px;
line-height: 24px;
margin-bottom: 0px;
margin-top: 0px;


<h2
  className="ub-mt_0px ub-mb_0px ub-color_101840 ub-fnt-fam_wm9v6k ub-f-wght_600 ub-fnt-sze_18px ub-ln-ht_24px ub-ltr-spc_-0-07px ub-box-szg_border-box"
>
  Heading 600
</h2>
`;

exports[`<Heading /> size 700 renders as expected 1`] = `
Extracted Styles:
box-sizing: border-box;
color: #101840;
font-family: "SF UI Display", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 20px;
font-weight: 600;
letter-spacing: -0.07px;
line-height: 24px;
margin-bottom: 0px;
margin-top: 0px;


<h2
  className="ub-mt_0px ub-mb_0px ub-color_101840 ub-fnt-fam_wm9v6k ub-f-wght_600 ub-fnt-sze_20px ub-ln-ht_24px ub-ltr-spc_-0-07px ub-box-szg_border-box"
>
  Heading 700
</h2>
`;

exports[`<Heading /> size 800 renders as expected 1`] = `
Extracted Styles:
box-sizing: border-box;
color: #101840;
font-family: "SF UI Display", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 24px;
font-weight: 600;
letter-spacing: -0.2px;
line-height: 32px;
margin-bottom: 0px;
margin-top: 0px;


<h2
  className="ub-mt_0px ub-mb_0px ub-color_101840 ub-fnt-fam_wm9v6k ub-f-wght_600 ub-fnt-sze_24px ub-ln-ht_32px ub-ltr-spc_-0-2px ub-box-szg_border-box"
>
  Heading 800
</h2>
`;

exports[`<Heading /> size 900 renders as expected 1`] = `
Extracted Styles:
box-sizing: border-box;
color: #101840;
font-family: "SF UI Display", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 32px;
font-weight: 600;
letter-spacing: -0.2px;
line-height: 40px;
margin-bottom: 0px;
margin-top: 0px;


<h2
  className="ub-mt_0px ub-mb_0px ub-color_101840 ub-fnt-fam_wm9v6k ub-f-wght_600 ub-fnt-sze_32px ub-ln-ht_40px ub-ltr-spc_-0-2px ub-box-szg_border-box"
>
  Heading 900
</h2>
`;
